pkg/kubectl/plugins/loader.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugins
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const PluginDescriptorFilename = "plugin.yaml"
+
+// PluginLoader is capable of loading a list of plugin descriptions.
+type PluginLoader interface {
+	Load() (Plugins, error)
+}
+
+// DirectoryPluginLoader is a PluginLoader that loads plugin descriptions
+// from a given directory in the filesystem. Plugins are located in subdirs
+// under the loader "root", where each subdir must contain, at least, a plugin
+// descriptor file called "plugin.yaml" that translates into a PluginDescription.
+type DirectoryPluginLoader struct {
+	Directory string
+}
+
+// Load reads the directory the loader holds and loads plugin descriptions.
+func (l *DirectoryPluginLoader) Load() (Plugins, error) {
+	if len(l.Directory) == 0 {
+		return nil, fmt.Errorf("Directory not specified")
+	}
+
+	list := Plugins{}
+
+	stat, err := os.Stat(l.Directory)
+	if err != nil {
+		return nil, err
+	}
+	if !stat.IsDir() {
+		return nil, fmt.Errorf("Not a directory: %s", l.Directory)
+	}
+
+	base, err := filepath.Abs(l.Directory)
+	if err != nil {
+		return nil, err
+	}
+
+	// read the base directory searching for subdirectories that may contain plugins
+	err = filepath.Walk(base, func(path string, fileInfo os.FileInfo, _ error) error {
+		if os.SameFile(stat, fileInfo) || !fileInfo.IsDir() {
+			return nil
+		}
+
+		// on each subdirectory, look for a 'plugin.yaml' descriptor
+		// fails silently (files unable to be read or unmarshalled are just skipped)
+		return filepath.Walk(path, func(item string, itemInfo os.FileInfo, _ error) error {
+			if itemInfo.IsDir() ||
+				itemInfo.Name() != PluginDescriptorFilename {
+				return nil
+			}
+
+			plugin := &Plugin{}
+			file, err := ioutil.ReadFile(item)
+			if err != nil {
+				glog.V(6).Infof("Unable to read plugin descriptor: %v", err)
+				return nil
+			}
+			if err := yaml.Unmarshal(file, plugin); err != nil {
+				glog.V(6).Infof("Unable to unmarshal plugin descriptor: %v", err)
+				return nil
+			}
+
+			plugin.Dir = path
```

Comment:
You need validation for the required fields. IIRC they won't error out during decoding, but will be empty and that might break your further parts of the code that assume its existence.